### PR TITLE
attempt to fix #5264

### LINKED ISSUE #5264: inheritance not work for generic object whose parent is parameterized
```nimrod
type
  abc[T,X] = object of T
    a: X
```

this simple type definition rejected by the compiler with following error message:
```text
inherit.nim(2, 21) Error: inheritance only works with non-final objects
```

surely this add to already long list of generic's type section bugs

### PR DESCRIPTION
**Pull request by jangko**

### CODE REVIEW

File: compiler/semtypes.nim [hunk 1]
```diff
@@ -1042,6 +1043,35 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
 proc semGenericParamInInvocation(c: PContext, n: PNode): PType =
   result = semTypeNode(c, n, nil)
 
+proc semRecordTypeAux(c: PContext, n: PNode, check: var IntSet) =
```
Comment (Araq) [inline compiler/semtypes.nim]: That looks like you copied it from somewhere. ;-) You can make the original a template perhaps and instantiate it here. Also your code is wrong ``case`` in objects can introduce different AST kinds (which is why the internal structure is so weird to begin with...).

**Comment by jangko** [comment]
I have modified semRecordNodeAux a little bit, removed the unnecessary semRecordTypeAux, and add more test covering object with case.
Is there any other things I missed?
Actually I discover another bug during investigation: #5521

File: compiler/semtypes.nim [hunk 2]
```diff
@@ -1104,6 +1129,14 @@ proc semGeneric(c: PContext, n: PNode, s: PSym, prev: PType): PType =
         result = instGenericContainer(c, n.info, result,
                                       allowMetaTypes = false)
 
+  # special check for generic object with
+  # parameterized parent
+  if result.kind == tyGenericInst:
+    var t = result.lastSon()
```
Comment (Araq) [inline compiler/semtypes.nim]: use something like this here:

```nim

let t = result.skipTypes(abstractPtrs)
if t != result and t.kind == tyObject and t.sons[0] != nil:
  semObjectType(c, n, t)
```

File: compiler/semtypes.nim [hunk 3]
```diff
@@ -1042,6 +1050,23 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
 proc semGenericParamInInvocation(c: PContext, n: PNode): PType =
   result = semTypeNode(c, n, nil)
 
+proc semObjectType(c: PContext, n: PNode, t: PType) =
```
Comment (Araq) [inline compiler/semtypes.nim]: name this ``semObjectTypeForInheritedGenericInst`` since it's only used for this case.

**Comment by Araq** [comment]
Looks good now, thanks! Will merge once AppVeyor has completed.

File: compiler/semtypes.nim [hunk 4]
```diff
@@ -1042,6 +1050,23 @@ proc semBlockType(c: PContext, n: PNode, prev: PType): PType =
 proc semGenericParamInInvocation(c: PContext, n: PNode): PType =
   result = semTypeNode(c, n, nil)
 
+proc semObjectTypeForInheritedGenericInst(c: PContext, n: PNode, t: PType) =
+  var check = initIntSet()
+  var realBase = t.sons[0]
+  var pos = 0
+  var base = skipTypesOrNil(realBase, skipPtrs)
```
Comment (Araq) [inline compiler/semtypes.nim]: I think ``realBase`` and ``base`` can be ``let`` here. No need to change it, just a friendly note for the next time: It's not necessary to copy the compiler's code style when it is a leftover from the Pascal days. :-)

Comment (jangko) [inline compiler/semtypes.nim]: 😄 👍

**Pull request merged**